In [54]:
import pandas as pd
from sklearn.model_selection import train_test_split
import time
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import chi2
from sklearn.feature_selection import RFE
from sklearn.linear_model import LogisticRegression
import pickle
import matplotlib.pyplot as plt

In [55]:
def split_scalar(indep_x,dep_Y):
    # Split the data into training and testing sets
    x_train,x_test,y_train,y_test=train_test_split(indep_x,dep_Y,test_size=0.30,random_state=0)
    sc = StandardScaler() 
    # the training data using the scaler "sc"
    x_train=sc.fit_transform(x_train)
    #Return training and testing sets to independent variables (x_train and x_test) adhoda training and testing sets for dependent variable (y_train and y_test):
    x_test=sc.transform(x_test)
    return x_train,x_test,y_train,y_test        

In [56]:
def selectKbest(indep_x, dep_Y, n):
    # Create an instance of SelectKBest with chi2 as the scoring function
    test = SelectKBest(score_func=chi2, k=n)
    # Fit the SelectKBest model using indep_x and dep_Y
    fit1 = test.fit(indep_x, dep_Y)
    # Transform indep_x using the fitted model
    selectK_features = fit1.transform(indep_x)
    # Return the transformed features
    return selectK_features

In [57]:
def split_scalar(indep_x,dep_Y):
    # Split the data into training and testing sets
    x_train,x_test,y_train,y_test=train_test_split(indep_x,dep_Y,test_size=0.30,random_state=0)
    sc = StandardScaler() 
    # the training data using the scaler "sc"
    x_train=sc.fit_transform(x_train)
    #Return training and testing sets to independent variables (x_train and x_test) adhoda training and testing sets for dependent variable (y_train and y_test):
    x_test=sc.transform(x_test)
    return x_train,x_test,y_train,y_test 

In [58]:
def cm_prediction(classifier,X_test):
        y_pred = classifier.predict(X_test)
        
        # Making the Confusion Matrix
        from sklearn.metrics import confusion_matrix
        cm = confusion_matrix(y_test, y_pred)
        from sklearn.metrics import accuracy_score 
        from sklearn.metrics import classification_report 
            #from sklearn.metrics import confusion_matrix
            #cm = confusion_matrix(y_test, y_pred)
        Accuracy=accuracy_score(y_test, y_pred )
        report=classification_report(y_test, y_pred)
        return  classifier,Accuracy,report,X_test,y_test,cm

In [59]:
def logistic(X_train,y_train,X_test):       
        # Fitting K-NN to the Training set
        from sklearn.linear_model import LogisticRegression
        classifier = LogisticRegression(random_state = 0)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm 

In [60]:
def svm_linear(X_train,y_train,X_test):              
        from sklearn.svm import SVC
        classifier = SVC(kernel = 'linear', random_state = 0)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm
    

In [61]:
def svm_NL(X_train,y_train,X_test):
                
        from sklearn.svm import SVC
        classifier = SVC(kernel = 'rbf', random_state = 0)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm

In [62]:
def Navie(X_train,y_train,X_test):       
        # Fitting K-NN to the Training set
        from sklearn.naive_bayes import GaussianNB
        classifier = GaussianNB()
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm  

In [63]:
def knn(X_train,y_train,X_test):           
        # Fitting K-NN to the Training set
        from sklearn.neighbors import KNeighborsClassifier
        classifier = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski', p = 2)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm

In [64]:
def Decision(X_train,y_train,X_test):        
        # Fitting K-NN to the Training set
        from sklearn.tree import DecisionTreeClassifier
        classifier = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm 

In [65]:
def random(X_train,y_train,X_test):        
        # Fitting K-NN to the Training set
        from sklearn.ensemble import RandomForestClassifier
        classifier = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 0)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm

In [66]:
def selectk_Classification(acclog,accsvml,accsvmnl,accknn,accnav,accdes,accrf):    
    dataframe=pd.DataFrame(index=['KernelPCA'],columns=['Logistic','SVMl','SVMnl','KNN','Navie','Decision','Random'])
    for number,idex in enumerate(dataframe.index):      
        dataframe['Logistic'][idex]=acclog[number]       
        dataframe['SVMl'][idex]=accsvml[number]
        dataframe['SVMnl'][idex]=accsvmnl[number]
        dataframe['KNN'][idex]=accknn[number]
        dataframe['Navie'][idex]=accnav[number]
        dataframe['Decision'][idex]=accdes[number]
        dataframe['Random'][idex]=accrf[number]
    return dataframe

In [68]:
dataset1=pd.read_csv("Car_Rating data.csv",index_col=None)

In [69]:
df2=dataset1

In [70]:
df2 = pd.get_dummies(df2, drop_first=True)

In [71]:
df2

Index  Overall Rating  Exterior  Interior  Ride Quality  Airbags  \
0        0             4.4       4.4       4.4           4.4       12   
1        1             4.2       4.0       3.8           4.2        8   
2        2             3.5       2.7       3.0           2.3        2   
3        3             4.4       4.4       4.4           4.4        0   
4        4             4.2       4.0       3.8           4.2        4   
..     ...             ...       ...       ...           ...      ...   
167    167             0.0       NaN       NaN           NaN        4   
168    168             0.0       NaN       NaN           NaN        8   
169    169             0.0       NaN       NaN           NaN        4   
170    170             0.0       NaN       NaN           NaN       12   
171    171             0.0       NaN       NaN           NaN        0   

     Price.1  Car_Audi A4  Car_Audi Q3  Car_BMW 2 Series Gran Coupe  ...  \
0      13328        False        False                        False  ...   
1      16621        False        False                        False  ...   
2       8467        False        False                        False  ...   
3       3607        False        False                        False  ...   
4      11726        False        False                        False  ...   
..       ...          ...          ...                          ...  ...   
167    20005        False        False                        False  ...   
168    36110        False        False                        False  ...   
169     7840        False        False                        False  ...   
170      470        False        False                        False  ...   
171     1300        False        False                        False  ...   

     Color_Red  Color_Silver  Color_Sky blue  Color_White  Color_Yellow  \
0        False          True           False        False         False   
1        False         False           False        False         False   
2        False         False           False        False         False   
3        False         False           False         True         False   
4        False          True           False        False         False   
..         ...           ...             ...          ...           ...   
167      False         False           False         True         False   
168      False          True           False        False         False   
169      False         False           False        False         False   
170      False         False           False        False         False   
171      False         False           False        False         False   

     Fuel type_Diesel  Fuel type_Hybrid  Fuel type_LPG  Fuel type_Petrol  \
0               False              True          False             False   
1               False             False          False              True   
2               False             False          False              True   
3               False              True          False             False   
4               False             False          False              True   
..                ...               ...            ...               ...   
167             False             False          False              True   
168             False             False          False              True   
169             False             False          False              True   
170             False             False          False              True   
171             False             False          False              True   

     Fuel type_Plug-in Hybrid  
0                       False  
1                       False  
2                       False  
3                       False  
4                       False  
..                        ...  
167                     False  
168                     False  
169                     False  
170                     False  
171                     False  

[172 rows x 241 

In [77]:
df2["Exterior"].fillna(0, inplace=True)
df2["Interior"].fillna(0, inplace=True)
df2["Ride Quality"].fillna(0, inplace=True)

In [78]:
indep_x=df2.drop('Price.1',axis=1)
dep_Y=df2['Price.1']

In [102]:
Kbest=selectKbest(indep_x,dep_Y,2)
acclog=[]
accsvml=[]
accsvmnl=[]
accknn=[]
accnav=[]
accdes=[]
accrf=[]

In [103]:
Kbest

array([[  0.,  12.],
       [  1.,   8.],
       [  2.,   2.],
       [  3.,   0.],
       [  4.,   4.],
       [  5.,   4.],
       [  6.,  12.],
       [  7.,  12.],
       [  8.,  12.],
       [  9.,  12.],
       [ 10.,  12.],
       [ 11.,   0.],
       [ 12.,   4.],
       [ 13.,  12.],
       [ 14.,   4.],
       [ 15.,  12.],
       [ 16.,  12.],
       [ 17.,   4.],
       [ 18.,  12.],
       [ 19.,  12.],
       [ 20.,   4.],
       [ 21.,  12.],
       [ 22.,  12.],
       [ 23.,   0.],
       [ 24.,   4.],
       [ 25.,   8.],
       [ 26.,   0.],
       [ 27.,   8.],
       [ 28.,   4.],
       [ 29.,  12.],
       [ 30.,   4.],
       [ 31.,  12.],
       [ 32.,   6.],
       [ 33.,  12.],
       [ 34.,  10.],
       [ 35.,   4.],
       [ 36.,  12.],
       [ 37.,   4.],
       [ 38.,   4.],
       [ 39.,  12.],
       [ 40.,   4.],
       [ 41.,   4.],
       [ 42.,   8.],
       [ 43.,  12.],
       [ 44.,  12.],
       [ 45.,  12.],
       [ 46.,   8.],
       [ 47.,

In [104]:
x_train,x_test,y_train,y_test = split_scalar(Kbest,dep_Y)
classifier,Accuracy,report,x_test,y_test,cm=logistic(x_train,y_train,x_test)
acclog.append(Accuracy)

classifier,Accuracy,report,x_test,y_test,cm=svm_linear(x_train,y_train,x_test)  
accsvml.append(Accuracy)
    
classifier,Accuracy,report,x_test,y_test,cm=svm_NL(x_train,y_train,x_test)  
accsvmnl.append(Accuracy)
    
classifier,Accuracy,report,x_test,y_test,cm=knn(x_train,y_train,x_test)  
accknn.append(Accuracy)
    
classifier,Accuracy,report,x_test,y_test,cm=Navie(x_train,y_train,x_test)  
accnav.append(Accuracy)
    
classifier,Accuracy,report,x_test,y_test,cm=Decision(x_train,y_train,x_test)  
accdes.append(Accuracy)
    
classifier,Accuracy,report,x_test,y_test,cm=random(x_train,y_train,x_test)  
accrf.append(Accuracy)
    
result=selectk_Classification(acclog,accsvml,accsvmnl,accknn,accnav,accdes,accrf)

C:\Users\ragus\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\ragus\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\ragus\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\ragus\anaconda3\Lib\site-packages\sklearn\metrics\_cla

In [85]:
result
#5

Logistic SVMl SVMnl       KNN     Navie Decision Random
KernelPCA      0.0  0.0   0.0  0.019231  0.076923      0.0    0.0

# **Naive Bayes** shows the highest performance with a score of 0.076923.
# - **KNN** has a score of 0.019231.
# - All other models (Logistic, SVM linear, SVM nonlinear, Decision Tree, Random Forest) have a score of 0.0.


In [89]:
result
#4

Logistic SVMl SVMnl       KNN     Navie  Decision Random
KernelPCA      0.0  0.0   0.0  0.038462  0.076923  0.019231    0.0

# **Naive Bayes** maintains the highest performance with a score of 0.076923.
# - **KNN** improves to a score of 0.038462.
# - **Decision Tree** also performs better with a score of 0.019231.
# - Other models (Logistic, SVM linear, SVM nonlinear, Random Forest) still have a score of 0.0.


In [93]:
result
#3

Logistic SVMl     SVMnl       KNN     Navie Decision Random
KernelPCA      0.0  0.0  0.019231  0.038462  0.076923      0.0    0.0

# Result #3:
# Naive Bayes maintains the highest performance with a score of 0.076923.
# KNN improves to a score of 0.038462.
# Other models (Logistic, SVM linear, Decision Tree, Random Forest) still have a score of 0.0.

In [105]:
result
#2

Logistic      SVMl     SVMnl       KNN     Navie  Decision    Random
KernelPCA      0.0  0.019231  0.038462  0.019231  0.096154  0.019231  0.019231

# Result #2:
# Naive Bayes maintains the highest performance with a score of 0.096154.
# KNN improves to a score of 0.019231.
# Decision Tree also performs better with a score of 0.019231.
# Other models (Logistic, SVM linear, SVM nonlinear, Random Forest) still have a score of 0.0.

In [101]:
result
#1

Logistic SVMl SVMnl       KNN     Navie Decision Random
KernelPCA      0.0  0.0   0.0  0.019231  0.019231      0.0    0.0